tests/guest-tools/win/test_guest_tools_win.py

```diff
@@ -0,0 +1,346 @@
+    def vm_install_xcpng_v8_tools(self, vm_instance: VM):
+        install_other_drivers(
+            vm_instance,
+            "xcp-ng-8.2.2.200\\managementagentx64.msi",
+            is_msi=True,
+        )
+        return vm_instance
+
+    @pytest.fixture
+    def vm_install_xcpng_v9_tools(self, vm_instance: VM):
+        vm = vm_instance
+
+        vm.insert_cd("other-guest-tools-win.iso")
+        time.sleep(5)
+
+        logging.info("Install VM root certs")
+        vm.ssh(
+            [
+                "certutil",
+                "-addstore",
+                "-f",
+                "Root",
+                "D:/xcp-ng-9.0.9000/testsign/XCP-ng_Test_Signer.crt",
+            ]
+        )
+        vm.ssh(
+            [
+                "certutil",
+                "-addstore",
+                "-f",
+                "TrustedPublisher",
+                "D:/xcp-ng-9.0.9000/testsign/XCP-ng_Test_Signer.crt",
+            ]
+        )
+
+        vm.eject_cd()
+
+        install_other_drivers(
+            vm,
+            "xcp-ng-9.0.9000\\package\\XenDrivers-x64.msi",
+            is_msi=True,
+        )
+        return vm
+
+    @pytest.fixture
+    def vm_install_wu_drivers(self, vm_prepared: VM):
+        (vm, snapshot) = vm_prepared
+        vm.param_set("has-vendor-device", True)
+        vm.start()
+        wait_for_guest_start(vm)
+        install_other_drivers(
+            vm,
+            "citrix-9.4.0",
+            is_msi=False,
+        )
+        yield vm
+        snapshot.revert()
+
+    def test_uninstall_tools(self, vm_install: VM):
+        vm = vm_install
+        vm.reboot()
+        wait_for_guest_start(vm)
+        logging.info("Uninstall Windows PV drivers")
+        uninstall_guest_tools(vm, action=PowerAction.Reboot)
+        assert not is_drivers_installed(vm)
+
+    def test_uninstall_tools_early(self, vm_install: VM):
+        vm = vm_install
+        logging.info("Uninstall Windows PV drivers before rebooting")
+        uninstall_guest_tools(vm, action=PowerAction.Reboot)
+        assert not is_drivers_installed(vm)
+
+    def test_reinstall_tools_early(self, vm_install: VM):
+        vm = vm_install
+        vm.reboot()
+        wait_for_guest_start(vm)
+        logging.info("Uninstall Windows PV drivers before reinstalling")
+        uninstall_guest_tools(vm, action=PowerAction.Nothing)
+        install_guest_tools(vm, action=PowerAction.Reboot, check=False)
+        assert is_drivers_installed(vm)
+
+    def test_install_with_citrix_tools(self, vm_install_citrix_tools: VM):
+        exitcode = install_cert_and_tools(
+            vm_install_citrix_tools, PowerAction.Nothing, check=False
+        )
+        assert exitcode == ERROR_INSTALL_FAILURE
+
+    def test_install_with_xcpng_v8_tools(self, vm_install_xcpng_v8_tools: VM):
+        exitcode = install_cert_and_tools(
+            vm_install_xcpng_v8_tools, PowerAction.Nothing, check=False
+        )
+        assert exitcode == ERROR_INSTALL_FAILURE
```

Comment:
Do you also envision tests where we start with either of the old v8 tools, xcpng_v9_tools, old citrix tools, recent citrix tools, remove them and replace them with the latest we built?

Reply:
Some tools (e.g. xcp-ng v8) don't uninstall cleanly and crash the VM. I'll try to add scenarios where some xenbus/etc. drivers are already installed (without the management packages).

Reply:
Could we use the clean-up tool? This would allow to test it in the process, too.